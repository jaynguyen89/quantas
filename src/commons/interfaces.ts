import { CancellationType, ImageType, PromotionType, RatingType } from 'src/commons/enums';

export interface IApiHotel {
    id: string,
    property: {
        propertyId: string,
        title: string,
        address: Array<string>,
        previewImage: {
            url: string,
            caption: string,
            imageType: string,
        },
        rating: {
            ratingValue: number,
            ratingType: string,
        },
    },
    offer: {
        promotion: {
            title: string,
            type: string,
        },
        name: string,
        displayPrice: {
            amount: number,
            currency: string,
        },
        savings: {
            amount: number,
            currency: string,
        },
        cancellationOption: {
            cancellationType: string,
        },
    },
}

export interface IHotel {
    id: string,
    name: string,
    address: string,
    image: {
        url: string,
        caption: string,
        type: ImageType,
    },
    rating: {
        value: number,
        type: RatingType,
    },
    pricing: IOffer,
}

export interface IOffer {
    name: string,
    price: {
        value: number,
        currency: string,
    },
    promotion: {
        title: string,
        type: PromotionType,
    },
    savings?: {
        value: number,
        currency: string,
    },
    cancellationType: CancellationType,
}