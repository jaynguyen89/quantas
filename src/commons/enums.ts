export enum ImageType {
    PRIMARY = 'PRIMARY',
    SECONDARY = 'SECONDARY',
}

export enum PromotionType {
    MEMBER = 'MEMBER',
    GUEST = 'GUEST',
}

export enum CancellationType {
    FREE_CANCELLATION = 'FREE_CANCELLATION',
    NOT_REFUNDABLE = 'NOT_REFUNDABLE',
}

export enum RatingType {
    SELF = 'SELF',
    STAR = 'STAR',
}

export enum SortingOption {
    HIGH_LOW = 'Price high-low',
    LOW_HIGH = 'Price low-high',
}