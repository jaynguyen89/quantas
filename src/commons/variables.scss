$baseSize: 10px;

// font weights
$thin: 100;
$light: 300;
$regular: 500;
$thick: 700;
$bold: 900;

// typography
$xh1: 8.0rem; // 80px
$mh1: 7.8rem; // 78px
$h1: 7.6rem; // 76px
$sh1: 7.4rem; // 74px
$xh2: 7.2rem; // 72px
$mh2: 7.0rem; // 70px
$h2: 6.8rem; // 68px
$sh2: 6.6rem; // 66px
$xh3: 6.4rem; // 64px
$mh3: 6.2rem; // 62px
$h3: 6.0rem; // 60px
$sh3: 5.8rem; // 58px
$xh4: 5.6rem; // 56px
$mh4: 5.4rem; // 54px
$h4: 5.2rem; // 52px
$sh4: 5.0rem; // 50px
$xh5: 4.8rem; // 48px
$mh5: 4.6rem; // 46px
$h5: 4.4rem; // 44px
$sh5: 4.2rem; // 42px
$xh6: 4.0rem; // 40px
$mh6: 3.8rem; // 38px
$h6: 3.6rem; // 36px
$sh6: 3.4rem; // 34px
$xh7: 3.2rem; // 32px
$mh7: 3.0rem; // 30px
$h7: 2.8rem; // 28px
$sh7: 2.6rem; // 26px
$xh8: 2.4rem; // 24px
$mh8: 2.2rem; // 22px
$h8: 2.0rem; // 20px
$sh8: 1.8rem; // 18px
$xh9: 1.6rem; // 16px
$mh9: 1.4rem; // 14px
$h9: 1.2rem; // 12px
$sh9: 1.0rem; // 10px
$xh10: 0.8rem; // 8px
$mh10: 0.6rem; // 6px
$h10: 0.4rem; // 4px
$sh10: 0.2rem; // 2px

// screen sizes
$ultrawide-width: 1536px;
$wide-width: 1280px;
$large-width: 1024px;
$desktop-width: 992px;
$medium-width: 667px;
$small-width: 480px;
$tiny-width: 360px;

// padding, margin, border, border radius
$xxlarge: 4.0rem; // 40px
$xlarge: 3.7rem; // 37px
$large: 3.4rem; // 34px
$xxmedium: 3.1rem; // 31px
$xmedium: 2.8rem; // 28px
$medium: 2.5rem; // 25px
$small: 2.2rem; // 22px
$xsmall: 1.9rem; // 19px
$xxsmall: 1.6rem; // 16px
$tiny: 1.3rem; // 13px
$xtiny: 1.0rem; // 10px
$xxtiny: 0.7rem; // 7px
$micro: 0.4rem; // 4px
$xmicro: 0.1rem; // 1px

// colors as in colors.ts
$white: #ffffff;
$black: #000000;
$lightGrey: #e1e1e1;
$grey: #9d9d9d;
$red: #9e293c;
$green: #466f5f;
$yellow: #f5e400;
$blue: #409bed;

$shadow: rgba(0, 0, 0, 0.24) 0px 3px 8px;