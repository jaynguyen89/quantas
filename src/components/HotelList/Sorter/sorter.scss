@use 'src/commons/variables' as vars;

.sorter {
    margin-top: vars.$micro;

    span {
        font-weight: vars.$thick;
        margin-right: vars.$micro;
    }

    select {
        letter-spacing: 0.5px;
        min-width: 130px;
        border: vars.$xmicro solid vars.$grey;
        border-radius: vars.$micro;
        padding: 2px vars.$micro;
        -webkit-appearance: none;
        -moz-appearance: none;
        background: transparent url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABYAAAAWCAYAAADEtGw7AAAACXBIWXMAAAsTAAALEwEAmpwYAAAA5ElEQVR4nGNwmP32Py0ww6jBDoMmKCJWvAdjqhpcvO3T/88///3/+uvf//Idn6hjcOPez////P0PB7///v/fsOczZQbX7Pr8/++//xgAZFHVrs/kGZy45sP/b7+wmAoFoGABqSHZ4OZ9n/+fefILLwapGTw5z3fRu/9p6z/ixSA1JBs84ejX/4RA/9EvpBvsOOft/xOPfuE09PSTX/+d55KZ3HwWvvt//dVvDEOvvfz933sh7mAgKvK8F777f/E5wvDzz37991qA31CiU4XL3Lf/N1z7AcYgNkWpwoFWyc1hxBkMAPX/ha6yMtRKAAAAAElFTkSuQmCC) no-repeat;
        background-position-x: 100%;
        background-position-y: -2px;
    }

    select::-ms-expand { display: none; }
}