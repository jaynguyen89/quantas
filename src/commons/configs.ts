import * as process from 'process';
import { SortingOption } from 'src/commons/enums';

const configs = {
    maxRatingCap: process.env.REACT_APP_MAX_RATING_CAP === undefined
        ? 5
        : +process.env.REACT_APP_MAX_RATING_CAP,
    defaultSortOption: (process.env.REACT_APP_DEFAULT_SORTING ?? 'HIGH_LOW') as keyof typeof SortingOption,
};

export default configs;