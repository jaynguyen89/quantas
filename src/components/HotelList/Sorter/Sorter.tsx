import React from 'react';
import 'src/components/HotelList/Sorter/sorter.scss';

type TSorterProps = {
    name: string,
};

const Sorter = ({
    name,
}: TSorterProps) => {
    return (
        <div className='sorter'>
            <span>Sort by</span>
            <select name={name}>
                <option value='1'>Price high-low</option>
                <option value='1'>Price low-high</option>
            </select>
        </div>
    );
};

export default Sorter;