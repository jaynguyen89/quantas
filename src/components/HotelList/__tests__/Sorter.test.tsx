import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from '@cfaester/enzyme-adapter-react-18';
import toJson from 'enzyme-to-json';
import { fireEvent, render } from '@testing-library/react';
import Sorter from 'src/components/HotelList/Sorter/Sorter';
import { SortingOption } from 'src/commons/enums';

configure({ adapter: new Adapter() });

describe('Sorter.tsx', () => {
    it('renders correctly', () => {
        const wrapper = shallow(
            <Sorter name='sorter' value='LOW_HIGH' handleChange={(val) => console.log(val)} />
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('contains necessary elements', () => {
        const wrapper = render(
            <Sorter name='sorter' value='LOW_HIGH' handleChange={(val) => console.log(val)} />
        );

        const sorterContainer = wrapper.container.getElementsByClassName('sorter');
        expect(sorterContainer.length).toBe(1);
        expect(sorterContainer[0]).toBeDefined();

        const labelEl = sorterContainer[0].getElementsByTagName('span');
        expect(labelEl.length).toBe(1);
        expect(labelEl[0]).toBeDefined();

        const selectEl = sorterContainer[0].getElementsByTagName('select');
        expect(selectEl.length).toBe(1);
        expect(selectEl[0]).toBeDefined();
        expect(selectEl[0].value).toEqual('LOW_HIGH');

        const optionEl = selectEl[0].getElementsByTagName('option');
        expect(optionEl.length).toBe(Object.keys(SortingOption).length);

        [...Array(Object.keys(SortingOption).length)].forEach((_, i) => {
            expect(optionEl[i].getAttribute('value')).toEqual(Object.keys(SortingOption)[i]);
            expect(optionEl[i].innerHTML).toEqual(Object.values(SortingOption)[i]);
        });
    });

    it('should handle selection', () => {
        const mockHandler = jest.fn().mockImplementation((val: string) => console.log(val));
        const consoleSpy = jest.spyOn(console, 'log');

        const wrapper = render(<Sorter name='sorter' value='LOW_HIGH' handleChange={mockHandler} />);
        const selectEl = wrapper.container.getElementsByTagName('select')[0];

        fireEvent.change(selectEl, {target: { value: 'HIGH_LOW' }});
        expect(mockHandler).toHaveBeenCalledTimes(1);
        expect(consoleSpy).toHaveBeenCalledWith('HIGH_LOW');
    });
});