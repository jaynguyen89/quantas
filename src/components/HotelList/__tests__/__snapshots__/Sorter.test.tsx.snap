// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sorter.tsx renders correctly 1`] = `
<div
  className="sorter"
>
  <span>
    Sort by
  </span>
  <select
    name="sorter"
    onChange={[Function]}
    value="LOW_HIGH"
  >
    <option
      key="HIGH_LOW"
      value="HIGH_LOW"
    >
      Price high-low
    </option>
    <option
      key="LOW_HIGH"
      value="LOW_HIGH"
    >
      Price low-high
    </option>
  </select>
</div>
`;
